File: app01_java_nft/backend/src/main/java/com/devpowers/nftcerts/certificate/CertificateService.java
package com.devpowers.nftcerts.certificate;

import com.devpowers.nftcerts.api.StoredUpload;
import com.devpowers.nftcerts.blockchain.ContractService;
import com.devpowers.nftcerts.blockchain.MintResult;
import com.devpowers.nftcerts.error.ArtworkNotFoundException;
import com.devpowers.nftcerts.error.ArtworkNotPinnedException;
import com.devpowers.nftcerts.error.CertificateNotFoundException;
import com.devpowers.nftcerts.error.DuplicateContentHashException;
import com.devpowers.nftcerts.error.NotVerifiedException;
import com.devpowers.nftcerts.error.RoyaltyOutOfRangeException;
import com.devpowers.nftcerts.identity.KycVerificationService;
import com.devpowers.nftcerts.storage.IpfsStorageService;
import com.devpowers.nftcerts.storage.PinResult;
import org.springframework.stereotype.Service;

import java.math.BigInteger;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Orchestrates the artwork -> IPFS pinning -> on-chain minting -> Certificate persistence
 * pipeline described in the platform's SDLC design docs.
 */
@Service
public class CertificateService {

    private final ArtworkRepository artworkRepository;
    private final CertificateRepository certificateRepository;
    private final IpfsStorageService ipfsStorageService;
    private final ContractService contractService;
    private final KycVerificationService kycVerificationService;

    public CertificateService(ArtworkRepository artworkRepository, CertificateRepository certificateRepository,
                               IpfsStorageService ipfsStorageService, ContractService contractService,
                               KycVerificationService kycVerificationService) {
        this.artworkRepository = artworkRepository;
        this.certificateRepository = certificateRepository;
        this.ipfsStorageService = ipfsStorageService;
        this.contractService = contractService;
        this.kycVerificationService = kycVerificationService;
    }

    /**
     * Creates and persists an {@link Artwork} from a previously-uploaded, hashed file, then pins
     * both the image and its generated metadata JSON to IPFS.
     */
    public Artwork createArtwork(StoredUpload upload, String title, String description, String medium,
                                  Integer yearCreated, Integer royaltyPercentageBps, String artistWalletAddress,
                                  String artistDid) {
        if (royaltyPercentageBps == null || royaltyPercentageBps < 0 || royaltyPercentageBps > 10000) {
            throw new RoyaltyOutOfRangeException(royaltyPercentageBps);
        }

        Artwork artwork = new Artwork(
                UUID.randomUUID(),
                UUID.randomUUID(),
                upload.originalFilename(),
                upload.sha256Hash(),
                title,
                description,
                medium,
                yearCreated,
                royaltyPercentageBps,
                artistWalletAddress,
                artistDid,
                ArtworkStatus.UPLOADED,
                Instant.now());

        PinResult imagePin = ipfsStorageService.pinFile(upload.content(), upload.originalFilename(), upload.contentType());
        artwork.setImageIpfsUri(imagePin.uri());

        Map<String, Object> metadata = buildMetadataPayload(artwork, imagePin);
        PinResult metadataPin = ipfsStorageService.pinJson(metadata, artwork.getId() + "-metadata");
        artwork.setMetadataIpfsUri(metadataPin.uri());

        artwork.setStatus(ArtworkStatus.PINNED);

        return artworkRepository.save(artwork);
    }

    private Map<String, Object> buildMetadataPayload(Artwork artwork, PinResult imagePin) {
        Map<String, Object> metadata = new LinkedHashMap<>();
        metadata.put("title", artwork.getTitle());
        metadata.put("description", artwork.getDescription());
        metadata.put("medium", artwork.getMedium());
        metadata.put("yearCreated", artwork.getYearCreated());
        metadata.put("artistDid", artwork.getArtistDid());
        metadata.put("sha256Hash", artwork.getSha256Hash());
        metadata.put("image", imagePin.uri());
        return metadata;
    }

    /**
     * Mints an on-chain certificate for a previously-pinned artwork, enforcing KYC verification
     * and duplicate-content-hash protection before sending the transaction.
     */
    public Certificate mintCertificate(UUID artworkId, String recipientAddress) {
        Artwork artwork = artworkRepository.findById(artworkId)
                .orElseThrow(() -> new ArtworkNotFoundException(artworkId));

        if (artwork.getStatus() != ArtworkStatus.PINNED) {
            throw new ArtworkNotPinnedException(artworkId);
        }

        if (!kycVerificationService.isVerified(artwork.getArtistWalletAddress())) {
            throw new NotVerifiedException(artwork.getArtistWalletAddress());
        }

        certificateRepository.findByContentHashHex(artwork.getSha256Hash())
                .ifPresent(existing -> {
                    throw new DuplicateContentHashException(artwork.getSha256Hash());
                });

        MintResult mintResult = contractService.mintCertificate(
                recipientAddress,
                artwork.getMetadataIpfsUri(),
                artwork.getSha256Hash(),
                artwork.getArtistWalletAddress(),
                BigInteger.valueOf(artwork.getRoyaltyPercentageBps()));

        Certificate certificate = new Certificate(
                mintResult.tokenId().longValueExact(),
                artwork,
                artwork.getSha256Hash(),
                mintResult.contractAddress(),
                mintResult.txHash(),
                recipientAddress,
                artwork.getRoyaltyPercentageBps(),
                artwork.getArtistWalletAddress(),
                Instant.now());

        artwork.setStatus(ArtworkStatus.MINTED);
        artworkRepository.save(artwork);

        return certificateRepository.save(certificate);
    }

    public Certificate getCertificate(Long tokenId) {
        return certificateRepository.findById(tokenId)
                .orElseThrow(() -> new CertificateNotFoundException(tokenId));
    }

    public List<Certificate> getCertificatesForWallet(String walletAddress) {
        return certificateRepository.findByOwnerAddress(walletAddress);
    }
}

File: app02_cpp_nft/backend/src/certificate/CertificateService.cpp
#include "certificate/CertificateService.h"

#include "error/Exceptions.h"
#include "util/Time.h"
#include "util/Uuid.h"

#include <json/json.h>

namespace nftcerts::certificate {

CertificateService::CertificateService(ArtworkRepository& artworkRepository,
                                        CertificateRepository& certificateRepository,
                                        storage::IpfsStorageService& ipfsStorageService,
                                        blockchain::ContractService& contractService,
                                        identity::KycVerificationService& kycVerificationService)
    : artworkRepository_(artworkRepository),
      certificateRepository_(certificateRepository),
      ipfsStorageService_(ipfsStorageService),
      contractService_(contractService),
      kycVerificationService_(kycVerificationService) {}

Artwork CertificateService::createArtwork(const StoredUpload& upload, const std::string& title,
                                           const std::string& description, const std::string& medium,
                                           std::optional<int> yearCreated, std::optional<int> royaltyPercentageBps,
                                           const std::string& artistWalletAddress, const std::string& artistDid) {
    if (!royaltyPercentageBps.has_value() || *royaltyPercentageBps < 0 || *royaltyPercentageBps > 10000) {
        throw error::RoyaltyOutOfRangeException(royaltyPercentageBps.value_or(-1));
    }

    Artwork artwork;
    artwork.id = util::randomUuid();
    artwork.fileId = util::randomUuid();
    artwork.originalFilename = upload.originalFilename;
    artwork.sha256Hash = upload.sha256Hash;
    artwork.title = title;
    artwork.description = description;
    artwork.medium = medium;
    artwork.yearCreated = yearCreated;
    artwork.royaltyPercentageBps = *royaltyPercentageBps;
    artwork.artistWalletAddress = artistWalletAddress;
    artwork.artistDid = artistDid;
    artwork.status = ArtworkStatus::UPLOADED;
    artwork.createdAt = util::nowIso8601();

    storage::PinResult imagePin = ipfsStorageService_.pinFile(upload.content, upload.originalFilename, upload.contentType);
    artwork.imageIpfsUri = imagePin.uri;

    Json::Value metadata;
    metadata["title"] = artwork.title;
    metadata["description"] = artwork.description;
    metadata["medium"] = artwork.medium;
    if (artwork.yearCreated.has_value()) {
        metadata["yearCreated"] = *artwork.yearCreated;
    } else {
        metadata["yearCreated"] = Json::nullValue;
    }
    metadata["artistDid"] = artwork.artistDid;
    metadata["sha256Hash"] = artwork.sha256Hash;
    metadata["image"] = imagePin.uri;

    storage::PinResult metadataPin = ipfsStorageService_.pinJson(metadata, artwork.id + "-metadata");
    artwork.metadataIpfsUri = metadataPin.uri;

    artwork.status = ArtworkStatus::PINNED;

    return artworkRepository_.save(artwork);
}

Certificate CertificateService::mintCertificate(const std::string& artworkId, const std::string& recipientAddress) {
    std::optional<Artwork> artworkOpt = artworkRepository_.findById(artworkId);
    if (!artworkOpt.has_value()) {
        throw error::ArtworkNotFoundException(artworkId);
    }
    Artwork artwork = *artworkOpt;

    if (artwork.status != ArtworkStatus::PINNED) {
        throw error::ArtworkNotPinnedException(artworkId);
    }

    if (!kycVerificationService_.isVerified(artwork.artistWalletAddress)) {
        throw error::NotVerifiedException(artwork.artistWalletAddress);
    }

    if (certificateRepository_.findByContentHashHex(artwork.sha256Hash).has_value()) {
        throw error::DuplicateContentHashException(artwork.sha256Hash);
    }

    blockchain::MintResult mintResult = contractService_.mintCertificate(
        recipientAddress, artwork.metadataIpfsUri, artwork.sha256Hash, artwork.artistWalletAddress,
        static_cast<uint64_t>(artwork.royaltyPercentageBps));

    Certificate certificate;
    certificate.tokenId = mintResult.tokenId;
    certificate.artwork = artwork;
    certificate.contentHashHex = artwork.sha256Hash;
    certificate.contractAddress = mintResult.contractAddress;
    certificate.txHash = mintResult.txHash;
    certificate.ownerAddress = recipientAddress;
    certificate.royaltyPercentageBps = artwork.royaltyPercentageBps;
    certificate.royaltyRecipient = artwork.artistWalletAddress;
    certificate.mintedAt = util::nowIso8601();

    artwork.status = ArtworkStatus::MINTED;
    artworkRepository_.save(artwork);

    return certificateRepository_.save(certificate);
}

Certificate CertificateService::getCertificate(uint64_t tokenId) {
    auto certificate = certificateRepository_.findById(tokenId);
    if (!certificate.has_value()) {
        throw error::CertificateNotFoundException(static_cast<long long>(tokenId));
    }
    return *certificate;
}

std::vector<Certificate> CertificateService::getCertificatesForWallet(const std::string& walletAddress) {
    return certificateRepository_.findByOwnerAddress(walletAddress);
}

}  // namespace nftcerts::certificate
